pandas/core/reshape/reshape.py

```diff
@@ -1089,6 +1089,198 @@ def get_empty_frame(data) -> DataFrame:
+                if prefix in col:
+                    variables_slice[prefix].append(col)
+    else:
+        sep_for_prefix = {}
+        variables_slice = {}
+        for col in data_to_decode.columns:
+            ps = [ps for ps in sep if ps in col]
+            if len(ps) == 0:
+                raise ValueError(
+                    f"Separator not specified for all columns; "
+                    f"First instance column: '{col}'"
+                )
+            prefix = col.split(ps[0])[0]
+            if prefix not in sep_for_prefix:
+                sep_for_prefix[prefix] = ps[0]
+            if prefix not in variables_slice:
+                variables_slice[prefix] = [col]
+            else:
+                variables_slice[prefix].append(col)
+        sep = sep_for_prefix
+
+    # validate number of dropped_first
+    def check_len(item, name) -> None:
+        if not len(item) == len(variables_slice):
+            len_msg = (
+                f"Length of '{name}' ({len(item)}) did not match the "
+                "length of the columns being encoded "
+                f"({len(variables_slice)})"
+            )
+            raise ValueError(len_msg)
+
+    if dropped_first:
+        if isinstance(dropped_first, dict):
+            check_len(dropped_first, "dropped_first")
+        elif isinstance(dropped_first, list):
+            check_len(dropped_first, "dropped_first")
+            dropped_first = dict(zip(variables_slice, dropped_first))
+        elif isinstance(dropped_first, str):
+            dropped_first = dict(
+                zip(variables_slice, [dropped_first] * len(variables_slice))
+            )
+        else:
+            raise TypeError(
+                f"Expected 'dropped_first' to be of type 'str', 'list', or 'dict'; "
+                f"Received 'dropped_first' of type: {type(dropped_first).__name__}"
+            )
+
+    cat_data = {}
+    for prefix, prefix_slice in variables_slice.items():
+        if sep is None:
+            cats = subset.copy()
+        else:
+            cats = [col[len(prefix + sep[prefix]) :] for col in prefix_slice]
+        assigned = data_to_decode[prefix_slice].sum(axis=1)
+        if any(assigned > 1):
+            raise ValueError(
+                f"Dummy DataFrame contains multi-assignment(s); "
+                f"First instance in row: {assigned.argmax()}"
+            )
+        elif any(assigned == 0):
+            if isinstance(dropped_first, dict):
+                cats.append(dropped_first[prefix])
+            else:
+                raise ValueError(
+                    f"Dummy DataFrame contains unassigned value(s); "
+                    f"First instance in row: {assigned.argmin()}"
+                )
+            data_slice = concat((data_to_decode[prefix_slice], assigned == 0), axis=1)
+        else:
+            data_slice = data_to_decode[prefix_slice]
+        cat_data[prefix] = data_slice.dot(np.array(cats))
+
+    return DataFrame(cat_data)
```

Comment:
Should it ever return a `Series`?  Not quite an inverse operation if it does not.

Reply:
Missed @jreback 's comment above.